cli/internal/transformerpipeline/pipeline.go

```diff
@@ -70,31 +71,49 @@ func (lp *TransformerPipeline) Send(data []byte) error {
 	}
 
 	if lp.clientWrappers[0].isClosed {
-		return errors.New("cannot send data to a closed pipeline")
+		return nil
 	}
 
-	return lp.clientWrappers[0].client.Send(&plugin.Transform_Request{Record: data})
+	sendCh := make(chan error)
+
+	// Send can block forever (e.g. if grpc buffer is full), so we run it asynchronously
+	// and check if pipeline is closed every second.
+	go func() {
+		err := lp.clientWrappers[0].client.Send(&plugin.Transform_Request{Record: data})
+		sendCh <- err
+	}()
+
+	select {
+	case err := <-sendCh:
+		return err
+	case <-time.After(1 * time.Second): // Check if pipeline is closed every second
+		if lp.clientWrappers[0].isClosed {
```

Comment:
There's still a data race here, the race detector catches it once in a while.

Reply:
You're right, I didn't even look into that; I was optimising for fixing the issue.

I replaced the `bool` with an `atomic.Bool` and now the races disappeared.